definitions/npm/styled-components_v2.x.x/flow_v0.53.x-/test_styled-components_v2.x.x.js

```diff
@@ -0,0 +1,260 @@
+import styled, {
+  ThemeProvider,
+  withTheme,
+  keyframes,
+  ServerStyleSheet,
+  StyleSheetManager
+} from 'styled-components'
+import React from 'react'
+import type {
+  Theme,
+  Interpolation,
+
+  // Temporary
+  ReactComponentFunctional,
+  ReactComponentClass,
+  ReactComponentStyled,
+  ReactComponentStyledTaggedTemplateLiteral,
+  ReactComponentUnion,
+  ReactComponentIntersection,
+} from 'styled-components'
+
+const TitleTaggedTemplateLiteral: ReactComponentStyledTaggedTemplateLiteral<{}> = styled.h1;
+
+const TitleStyled: ReactComponentStyled<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+const TitleGeneric: ReactComponentIntersection<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+const TitleFunctional: ReactComponentFunctional<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+const TitleClass: ReactComponentClass<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+declare var needsReactComponentFunctional: ReactComponentFunctional<*> => void
+declare var needsReactComponentClass: ReactComponentClass<*> => void
+
+needsReactComponentFunctional(TitleStyled)
+needsReactComponentClass(TitleStyled)
+
+const ExtendedTitle: ReactComponentIntersection<*> = styled(TitleStyled)`
+  font-size: 2em;
+`;
+
+const Wrapper: ReactComponentIntersection<*> = styled.section`
+  padding: 4em;
+  background: ${({theme}) => theme.background};
+`;
+
+// ---- EXTEND ----
+const Attrs0ReactComponent: ReactComponentStyled<*> = styled.div.extend``;
+const Attrs0ExtendReactComponent: ReactComponentIntersection<*> = Attrs0ReactComponent.extend``;
+
+const Attrs0SyledComponent: ReactComponentStyledTaggedTemplateLiteral<*> = styled.div;
+const Attrs0ExtendStyledComponent: ReactComponentIntersection<*> = Attrs0SyledComponent.extend``;
+
+// ---- ATTRIBUTES ----
+const Attrs1: ReactComponentStyledTaggedTemplateLiteral<*> = styled.section.attrs({
+  testProp: 'foo'
+});
+
+// $ExpectError
+const Attrs1Error: ReactComponentStyledTaggedTemplateLiteral<*> = styled.section.attrs({
+  testProp: 'foo'
+})``;
+
+declare var needsString: string => void
+
+needsReactComponentFunctional(styled.section.attrs({})``)
+needsReactComponentClass(styled.section.attrs({})``)
+
+// $ExpectError
+needsString(styled.section.attrs({})``)
+
+const Attrs2: ReactComponentStyledTaggedTemplateLiteral<*> = styled.section
+  .attrs({
+    testProp1: 'foo'
+  })
+  .attrs({
+    testProp2: 'bar'
+  });
+
+const Attrs3Styled: ReactComponentStyled<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const Attrs3Generic: ReactComponentIntersection<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const Attrs3Functional: ReactComponentFunctional<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const Attrs3Class: ReactComponentClass<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const theme: Theme = {
+  background: "papayawhip"
+};
+
+// ---- WithComponent ----
```

Comment:
Basically I ended up with [this definition (see gist)](https://gist.github.com/michalkvasnicak/5bd086562487bd3a6c72b2bc078442e4) and it is working just fine. Only thing is that I have to typecast a return type like this `const A: typeof Component = styled(Component).withComponent('a')`.